### PR TITLE
WELD-2120 don't notify AFTER_SUCCES transactional observer when

### PR DESCRIPTION
**Pull request by tremes**
transaction was rollbacked

### CODE REVIEW

File: tests-arquillian/src/test/java/org/jboss/weld/tests/event/observer/transactional/rollback/TransactionalObserverRollbackTest.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+                .addClass(ActionSequence.class)
+                .addAsWebInfResource(EmptyAsset.INSTANCE, "beans.xml")
+                .addAsResource(TransactionalObserverRollbackTest.class.getPackage(), "persistence.xml", "META-INF/persistence.xml");
+    }
+
+    @Test
+    public void afterSuccessObserverIsNotNotifiedAfterTxRollBack() throws SystemException {
+        ActionSequence.reset();
+        ejbTestBean.initTransaction();
+        Assert.assertFalse(ActionSequence.getSequenceData().isEmpty());
```
Comment (mkouba) [inline tests-arquillian/src/test/java/org/jboss/weld/tests/event/observer/transactional/rollback/TransactionalObserverRollbackTest.java]: Shouldn't we rather compare the whole sequence data?

Comment (tremes) [inline tests-arquillian/src/test/java/org/jboss/weld/tests/event/observer/transactional/rollback/TransactionalObserverRollbackTest.java]: Probably yes..

File: modules/jta/src/main/java/org/jboss/weld/jta/TransactionalObserverNotifier.java [hunk 2]
```diff
@@ -78,7 +82,19 @@ public ObserverNotifier create(String contextId, TypeSafeObserverResolver resolv
             for (ObserverMethod<? super T> observer : observers) {
                 deferNotification(event, metadata, observer, notifications);
             }
-            transactionServices.registerSynchronization(new TransactionNotificationSynchronization(notifications));
+            try {
+                transactionServices.registerSynchronization(new TransactionNotificationSynchronization(notifications));
+            } catch (Exception e) {
+                if (e.getCause() instanceof RollbackException || e.getCause() instanceof IllegalStateException) {
+                    List<ObserverMethod<? super T>> filteredObservers = observers.stream()
```
Comment (mkouba) [inline modules/jta/src/main/java/org/jboss/weld/jta/TransactionalObserverNotifier.java]: We should sort the observers as desribed in https://github.com/cdi-spec/cdi/pull/272#issuecomment-205210486

**Comment by mkouba** [comment]
Otherwise looks good, please send a pull request for 2.3 branch as well.

File: modules/jta/src/main/java/org/jboss/weld/jta/TransactionalObserverNotifier.java [hunk 3]
```diff
@@ -78,7 +82,22 @@ public ObserverNotifier create(String contextId, TypeSafeObserverResolver resolv
...
+                            .filter(observerMethod -> !observerMethod.getTransactionPhase().equals(TransactionPhase.AFTER_SUCCESS)).collect(
+                                    Collectors.toList());
+                    filteredObservers.sort((o1, o2) -> {
```
Comment (mkouba) [inline modules/jta/src/main/java/org/jboss/weld/jta/TransactionalObserverNotifier.java]: `java.util.stream.Stream.sorted(Comparator<? super T>)` should work here...

Comment (tremes) [inline modules/jta/src/main/java/org/jboss/weld/jta/TransactionalObserverNotifier.java]: Yes but I guess I need to use Collector again then. This seemed to me more straightforward.

Comment (mkouba) [inline modules/jta/src/main/java/org/jboss/weld/jta/TransactionalObserverNotifier.java]: Hm, I think this should work: `.filter(observerMethod -> !observerMethod.getTransactionPhase().equals(TransactionPhase.AFTER_SUCCESS))
.sorted((o1, o2) -> o1.getTransactionPhase().compareTo(o2.getTransactionPhase())
.collect(Collectors.toList())`

But I don't like the comparison by ordinals... https://github.com/weld/core/pull/1315#discussion_r58501527

**Comment by tremes** [comment]
retest this please

**Comment by mkouba** [comment]
Merged, thanks!

**Pull request closed without merging**